versioned_docs/version-4.0.0-alpha.1/datastructures/storage-layout.md

```diff
@@ -0,0 +1,160 @@
+---
+title: Storage Layout
+slug: /datastructures/storage-layout
+---
+
+Smart contract authors are given some flexibility in regards on how they want to organize 
+the storage layout of their contracts.
+Let's dive deeper into the concepts behind ink! storage to get a better understanding 
+of some of its implications and limitations.
+
+## Storage Organization
+
+The following schema depicts the storage which is exposed
+to ink! by the contracts pallet:
+
+<div class="schema">
+    <img src="/img/kv.svg" alt="Storage Organization: Layout" />
+</div>
+
+Storage data is always encoded with the 
+[`SCALE`](https://docs.substrate.io/reference/scale-codec/) codec. 
+The storage API operates by storing and loading entries into and from a single storages 
+cells, where each storage cell is accessed under its own dedicated storage key. To some 
+extent, the storage API works similar to a traditional key-value database.
+
+## Packed vs Non-Packed layout
+
+Types that can be stored entirely under a single storage cell are considered
+[`Packed`](https://docs.rs/ink_storage_traits/4.0.0-beta.1/ink_storage_traits/trait.Packed.html).
+By default, ink! tries to store all storage struct fields under a single storage cell.
+Consequentially, with a `Packed` storage layout, any message interacting with the contract 
+storage will always need to operate on the entire contract storage struct.
+
+For example, if we have a somewhat small contract storage struct consisting of only a few 
+tiny fields, pulling everything from the storage inside every message is not 
+problematic. It may even be advantegous - especially if we expect most messages to 
+interact with most of the storage fields.
+
+On the other hand, this can get problematic if we're storing a large `ink::prelude::vec::Vec`
+in the contract storage but provide messages that do not need to read and write from this 
+`ink::prelude::vec::Vec`. In that scenario, each and every contract message bears runtime 
+overhead by dealing with that `ink::prelude::vec::Vec`, regardless whether they access it or 
+not. This results in extra gas costs.  To solve this problem we need to turn our storage 
+into a non-packed layout somehow.
```

Comment:
I'd only use the full import path once and then refer to it as `Vec` in later parts where its clear from the context that we're referring to the same `Vec`